src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs

```diff
@@ -188,28 +192,22 @@ private void ValidateService(ServiceDescriptor descriptor)
                 if (callSite.Cache.Location == CallSiteResultCacheLocation.Root)
                 {
                     object? value = CallSiteRuntimeResolver.Instance.Resolve(callSite, Root);
-                    return scope =>
-                    {
-                        DependencyInjectionEventSource.Log.ServiceResolved(this, serviceIdentifier.ServiceType);
-                        return value;
-                    };
+                    return new ServiceAccessor { CallSite = callSite, RealizedService = scope => value };
```

Comment:
So the issue was that the CallSite was not captured previously?

Reply:
Unfortunately, I don't quite understand why this caused the error. At the moment I can't write a unit test that reproduces the crash. I just rolled back the logic as it was and tried it on ef core tests. The tests were green, at least on my machine. More research needed

Reply:
I have a local repro. I'll update this PR with that soon.

Reply:
I wasn't able to push to your branch - see the comments on the PR for the test.

The issue is related to swapping to an emit-based version of the service accessor which doesn't have the logic that was added to the previous lambda.